src/Console/RecomposeCommand.php

```diff
@@ -0,0 +1,159 @@
+                new InputOption(
+                    'strict',
+                    'S',
+                    InputOption::VALUE_NONE,
+                    'Prefer ~ to ^ avoid accidental updates'
+                ),
+                new InputOption(
+                    'recipe-core-constraint',
+                    'R',
+                    InputOption::VALUE_OPTIONAL,
+                    'Version of `silverstripe/recipe-core` you are targeting. Defaults to the last stable',
+                    '*'
+                ),
+                new InputOption(
+                    'composer-path',
+                    'P',
+                    InputOption::VALUE_OPTIONAL,
+                    'Prefer ~ to ^ avoid accidental upgrades.',
+                    ''
+                )
+            ]);
+    }
+
+    /**
+     * @param InputInterface $input
+     * @param OutputInterface $output
+     * @return int|null
+     */
+    protected function execute(InputInterface $input, OutputInterface $output)
+    {
+        // Get our input variables
+        $rootPath = $this->getRootPath($input);
+        $write = $input->getOption('write');
+
+        $composerPath = $input->getOption('composer-path');
+        $recipeCoreConstraint = $input->getOption('recipe-core-constraint');
+        $coreTarget = $this->findTargetRecipeCore($recipeCoreConstraint);
+        $strict = $input->getOption('strict');
+
+        // Initialise our composer file.
+        $composer = new ComposerExec($rootPath, $composerPath);
+        $schema = new ComposerFile($composer, $rootPath);
+
+        // Set up some caching
+        $this->initPackageCache($composer);
+
+        // Set up our rules
+        $rules = [
+            new Rules\PhpVersion(),
+             new Rules\Rebuild($coreTarget),
+        ];
+        if ($strict) {
+            $rules[] = new Rules\StrictVersion();
+        }
+
+        // Try to upgrade the project
+        $change = $schema->upgrade($rules);
+
+        // Check if we got new content
+        if (!$change->hasNewContents($schema->getFullPath())) {
+            $output->writeln("Nothing to upgrade.");
+            return null;
+        }
+        $display = new ChangeDisplayer();
+        $display->displayChanges($output, $change);
+
+        // Ask the use if they want to save their changes.
+        // This is not the standard behavior, but this command can take quite a bit of time to run.
+        if (!$write) {
+            $helper = $this->getHelper('question');
+            $question = new ConfirmationQuestion('Save changes to composer file? (y/N)', false);
+            $write = $helper->ask($input, $output, $question);
+        }
+
+        if ($write) {
+            $schema->setContents($change->newContents($schema->getFullPath()));
+            $output->writeln("Changes have been saved.");
+        } else {
+            $output->writeln("Changes not saved; Run with --write to commit to disk");
+        }
+
+
+        return null;
```

Comment:
As it says in the docs, this function takes a _long_ time to run. Can we put some more output into it? A little bit of feedback would go a long way toward giving the user peace of mind that things are happening as intended.

Reply:
What about I update the output so it say something like this?
```
* Upgrading PHP constraint
* Rebuilding dependencies
  * Requiring PHP ... ✅
  * Requiring ext-json ...  ✅
  * Requiring silverstripe/recipe-core ... ✅
  * Requiring silverstripe/recipe-cms ... ✅
  * Requiring old-ss3/package-without-ss4 ... ❎
  * Requiring silverstripe/contentreview ... ✅
  * Requiring silverstripe/sharedraftcontent ... ✅
  * Requiring symbiote/silverstripe-advancedworkflow ...✅
  * Manually re-adding failed dependencies
* Switching to recipes
 * Switching to silverstripe/recipe-collaboration
   * Removing silverstripe/recipe-cms
   * Removing silverstripe/contentreview
   * Removing silverstripe/sharedraftcontent
   * Removing symbiote/silverstripe-advancedworkflow
* Using strict version constraint
* Showing difference
... diff here ...

Some of your dependencies could not be resolved. You'll need to manually correct those failed dependencies and run a `composer update`.
```